marklogic-data-hub/src/main/resources/ml-modules/root/marklogic-data-hub-spark-connector/writeLib.sjs

```diff
@@ -77,9 +77,34 @@ function buildMetadata(endpointConstants) {
   return metadata;
 }
 
-function generateUri(record, endpointConstants) {
-  const uriPrefix = endpointConstants.uriprefix != null ? endpointConstants.uriprefix : "";
-  return uriPrefix + sem.uuidString() + ".json";
+/**
+ * Generate an uri for an input object.
+ *
+ * @param uriTemplateRegEx The precompiled URI regular expression or null if not used.
+ * @param record The record we want to create an URI for.
+ * @param endpointConstants The input options
+ * @returns A random or template generated URI based on the options.
+ */
+function generateUri(uriTemplateRegEx,record, endpointConstants) {
+  if ( uriTemplateRegEx === null ) {
+    const uriPrefix = endpointConstants.uriprefix != null ? endpointConstants.uriprefix : "";
+    return uriPrefix + sem.uuidString() + ".json";
+  } else {
+    return endpointConstants.uritemplate.replace(uriTemplateRegEx, function (match, group) {
+      let propertyValue = record[group];
+      const type = typeof propertyValue;
+      if ( propertyValue === null ) {
+        throw new Error("Property ["+group+"] is null. This can't be used in a uri template.");
```

Comment:
For the error messages, we want to have strict control over the HTTP response code. So import ds-util.sjs and do the following:

```
ds.throwBadRequest("uriTemplate " + uriTemplate + " cannot be applied; property '" + group + "' is null");
```

Reply:
ready for commit